### DISCUSSION TITLE
Deprecate APIs to handle private key from headless

### DISCUSSION THREAD

**Discussion by moreal** [planetarium/NineChronicles.Standalone#1567] [General]
## 💭 Motivation

Currently, there is `ActionMutation` in GraphQL schema. The mutation makes a transaction with chosen action and stage the transaction automatically. To use the feature, you should make the headless have your private key through `/set-private-key` HTTP request or `--private-key` option. When everyone runs the headless locally, we made the mutation and used the feature in the launcher also. But the launcher became to make the actions through CLI tool and sign the transaction in local because it became default to use RPC mode.

I think it cannot be recommended to handle and transfer private key in raw. So I want to obsolete the below APIs:

- `mutation.aciton` `ActionMutation` GraphQL API
- `query.keyStore` `KeyStoreType` GraphQL API
- `mutation.keyStore` `KeyStoreMutation` GraphQL API
- `/set-private-key` HTTP API

## ❓ Migration

There are some ways to migrate your code using `ActionMutation`:

- With `actionQuery` GraphQL query. (See https://devforum.nine-chronicles.com/t/transfer-asset-with-graphql-queries/59/3)
- With `NineChronicles.Headless.Executable action` command. (Currently it's used by launcher).
- With Lib9c (If you write C# application)
- [WIP] With [planetarium/lib9c-wasm](https://github.com/planetarium/lib9c-wasm) (CURRENTLY NOT OFFICIAL, @moreal's personal experimental project to serve lib9c actions as JavaScript API through WASM)

## 📆 Plans (not fixed)

- At the v100310, the APIs will be deprecated. The GraphQL APIs will have `@deprecated` directives but you can use them yet. But the launcher will not call `/set-private-key` ([current](https://github.com/planetarium/9c-launcher/blob/aed7f5c20e176535f9ad248cb4711b2490d6c17b/src/v2/views/LoginView/index.tsx#L65)) so you must call `/set-private-key` with your private key if you want to use the deprecated GraphQL APIs though we don't recommend to use it.
  - [x] https://github.com/planetarium/NineChronicles.Headless/pull/1568
  - [ ] https://github.com/planetarium/9c-launcher/issues/1940
- About the v100350, the APIs will be obsoleted. (When we provide a way to migrate)

----

If you have opinions or questions for this plan, please leave as comments.

**Comment by BasixKOR** [comment]
We have #1559 for crafting a transaction, but it doesn't handle signing (by design). We're actively looking into solutions to this.
